src/app/icd/client/DefaultCheckInDelegate.cpp

```diff
@@ -44,5 +44,44 @@ void DefaultCheckInDelegate::OnCheckInComplete(const ICDClientInfo & clientInfo)
 #endif
 }
 
+ICDRefreshKeyInfo * DefaultCheckInDelegate::OnKeyRefreshNeeded(const ICDClientInfo & clientInfo,
+                                                               const ICDClientStorage * clientStorage)
+{
+    auto icdRefreshKeyInfo = CreateICDRefreshKeyInfo(clientInfo, clientStorage);
+
+    if (icdRefreshKeyInfo == nullptr)
+    {
+        return nullptr;
+    }
+
+    CHIP_ERROR err = chip::Crypto::DRBG_get_bytes(icdRefreshKeyInfo->mNewKey.Bytes(), icdRefreshKeyInfo->mNewKey.Capacity());
+    if (err != CHIP_NO_ERROR)
+    {
+        return nullptr;
+    }
+
+    return icdRefreshKeyInfo;
+}
+
+void DefaultCheckInDelegate::OnRegistrationUpdateComplete(const ICDClientInfo & clientInfo, ICDRefreshKeyInfo * icdRefreshKeyInfo)
+{
+    ChipLogProgress(ICD, "Re-registration with new key successful. New start counter = %d", clientInfo.start_icd_counter);
+    ReleaseICDRefreshKeyInfo(icdRefreshKeyInfo);
+}
+void DefaultCheckInDelegate::OnRegistrationUpdateFailure(ICDRefreshKeyInfo * icdRefreshKeyInfo)
+{
+    ReleaseICDRefreshKeyInfo(icdRefreshKeyInfo);
+}
+
+ICDRefreshKeyInfo * DefaultCheckInDelegate::CreateICDRefreshKeyInfo(const ICDClientInfo & clientInfo,
+                                                                    const ICDClientStorage * clientStorage)
+{
+    return chip::Platform::MakeUnique<ICDRefreshKeyInfo>(this, clientInfo, clientStorage).get();
```

Comment:
Assume you use adopt to pass the refreshKeyinfo ownership to callback, when on response or on error or is called, you delete callback, in destructor of callback, you mark RefreshInfo as null, and free when it is out of scope.
You can store refreshKeyInfo unique pointer in callback, when callback is out of scope, refreshInfo is freed.
Please refactor onResponse/OnError...to dedicated callback clas inside ICDRefreshKeyInfo, thanks

Reply:
We don't need transfers with the new design.